django/db/models/functions/text.py

```diff
@@ -65,83 +68,59 @@ def as_sqlite(self, compiler, connection, **extra_context):
         return super().as_sql(compiler, connection, function="CHAR", **extra_context)
 
 
-class ConcatPair(Func):
+class Concat(Func):
     """
-    Concatenate two arguments together. This is used by `Concat` because not
-    all backend databases support more than two arguments.
+    Concatenate expressions together into a string.
+
+    Handle nulls and non-textual expressions gracefully.
     """
 
-    function = "CONCAT"
+    function = None
+    template = "%(expressions)s"
+    arg_joiner = " || "
 
-    def as_sqlite(self, compiler, connection, **extra_context):
-        coalesced = self.coalesce()
-        return super(ConcatPair, coalesced).as_sql(
-            compiler,
-            connection,
-            template="%(expressions)s",
-            arg_joiner=" || ",
-            **extra_context,
+    def __init__(self, *expressions, **extra):
+        if len(expressions) < 2:
+            raise ValueError("Concat must take at least two expressions")
+        super().__init__(*expressions, **extra)
+
+    def as_mysql(self, compiler, connection):
+        # MySQL || operator is an alias for OR unless PIPES_AS_CONCAT is enabled.
+        return Func(Value(""), *self.source_expressions, function="CONCAT_WS").as_sql(
+            compiler, connection
         )
 
-    def as_postgresql(self, compiler, connection, **extra_context):
+    def as_sql(self, compiler, connection, **extra_context):
         copy = self.copy()
-        copy.set_source_expressions(
-            [
-                Cast(expression, TextField())
-                for expression in copy.get_source_expressions()
+        source_expressions = copy.get_source_expressions()
+        if connection.features.requires_casted_text_in_pipes_concat:
+            source_expressions = [
+                expr
+                if isinstance(expr.output_field, (CharField, TextField))
```

Comment:
I would add here also all the other text based fields (e.g. EmailField, SlugField, ...) to avoid an unnecessary text cast on them.

Reply:
EmailField & SlugField are CharFields though 🤔 The reason why Simon's put both Char & Text is because they independently inherit from Field.

Reply:
@pauloxnet exactly what @shangxiao said, `isinstance` respects inheritance rules and all fields used to store strings that Django provides are subclasses of either.

Reply:
@charettes thanks for the explanation. 😃 

However, I have verified that in addition to the fields that inherit from `CharField` and `TextField`, there are also other fields that are stored on "varchar" _(or similar)_ type columns in `PostgreSQL` and `SQLite` (`FilePathField`, `FileField`, `ImageField`) and still others if we consider only `SQLite` (`UUIDField`, `GenericIPAddressField`, `JSONField`). _(I haven't checked in `MariaDB`, `MySQL` and `Oracle`)_. Furthermore, a `ForeignKeyField` could also be of char/text type if the `to_field` attribute points to a text/char field in the referenced object. _(I haven't checked for the other relationship fields `ManyToManyField` and `OneToOneField`)_

Reply:
@pauloxnet the goal here is not to avoid casting at all cost, just like we don't avoid coalescing at all cost. We avoid casting when we're entirely sure that it's not necessary and not the other way around.

It's true that [casts are not necessary for uuid and inet on Postgres](https://dbfiddle.uk/mh_7rOhv) but they are not harmful either.

The goal of this PR is to have `Concat` generate SQL that can be used in context where `IMMUTABLE` is required while supporting passing nullable non-integer fields reference without explicit casts from the user (as that has been the _contract_ that `Concat` support over the years). It is *not* to generate the less verbose SQL for all backends that are supported at the cost of introducing complexity or abstraction layer leaks (e.g. such as introspecting the database specific column being generated).

Reply:
> @pauloxnet the goal here is not to avoid casting at all cost ...

Thanks. The goal is clear. 👍